contracts/examples/VCPayment-v2.sol

```diff
@@ -0,0 +1,228 @@
+        if (ownerPartPercent > 0 && ownerPartPercent < 100) {
+            revert InvalidOwnerPartPercent('Invalid owner part percent');
+        }
+        ownerPartPercent = _ownerPartPercent;
+    }
+
+    function setPaymentValue(
+        uint256 issuerId,
+        uint256 schemaHash,
+        uint256 value,
+        address withdrawAddress
+    ) public onlyOwner {
+        if (withdrawAddress == address(0)) {
+            revert InvalidWithdrawAddress('Invalid withdraw address');
+        }
+        PaymentData memory newPaymentData = PaymentData(
+            issuerId,
+            schemaHash,
+            value,
+            withdrawAddress,
+            0
+        );
+        paymentDataIds.push(keccak256(abi.encode(issuerId, schemaHash)));
+        _setPaymentData(issuerId, schemaHash, newPaymentData);
+    }
+
+    function updateWithdrawAddress(
+        uint256 issuerId,
+        uint256 schemaHash,
+        address withdrawAddress
+    ) external ownerOrIssuer(issuerId, schemaHash) {
+        if (withdrawAddress == address(0)) {
+            revert InvalidWithdrawAddress('Invalid withdraw address');
+        }
+        PaymentData memory payData = getPaymentData(issuerId, schemaHash);
+        uint256 issuerBalance = issuerAddressBalance[payData.withdrawAddress];
+        issuerAddressBalance[payData.withdrawAddress] = 0;
+        issuerAddressBalance[withdrawAddress] = issuerBalance;
+
+        payData.withdrawAddress = withdrawAddress;
+        _setPaymentData(issuerId, schemaHash, payData);
+    }
+
+    function updateValueToPay(
+        uint256 issuerId,
+        uint256 schemaHash,
+        uint256 value
+    ) external ownerOrIssuer(issuerId, schemaHash) {
+        PaymentData memory payData = getPaymentData(issuerId, schemaHash);
+        payData.valueToPay = value;
+        _setPaymentData(issuerId, schemaHash, payData);
+    }
+
+    function pay(string calldata paymentId, uint256 issuerId, uint256 schemaHash) external payable {
+        bytes32 payment = keccak256(abi.encode(issuerId, paymentId));
+        if (payments[payment]) {
+            revert PaymentError('Payment already done');
+        }
+        PaymentData memory payData = paymentData[keccak256(abi.encode(issuerId, schemaHash))];
+        if (payData.valueToPay == 0) {
+            revert PaymentError('Payment value not found for this issuer and schema');
+        }
+        if (payData.valueToPay != msg.value) {
+            revert PaymentError('Invalid value');
+        }
+        payments[payment] = true;
+
+        uint256 ownerPart = (msg.value * ownerPartPercent) / 100;
+        uint256 issuerPart = msg.value - ownerPart;
+
+        issuerAddressBalance[payData.withdrawAddress] += issuerPart;
+        ownerBalance += ownerPart;
+
+        payData.totalValue += issuerPart;
+        _setPaymentData(issuerId, schemaHash, payData);
+        emit Payment(issuerId, paymentId, schemaHash, block.timestamp);
```

Comment:
You don't need the block.timestamp in the event. It's exists for every event anyway at protocol level.

Reply:
removed